### ISSUE TITLE
idprovider doesn't have any site context, even with virtualhost on

### ISSUE THREAD

**Issue by hjelmevold** [enonic/xp-apps#334]
An idprovider only gets a very limited context, crucially not having any information about the site that the request originated from. The request object itself does contain the URL and servername and so forth, but not any site ID or anything that would allow me to receive data from the site or siteConfig.

It seems that this also is the case even when the site has been mapped in the vhost.cfg

Below is all the information that the get() function in the context lib returns:

```
{
    "branch": "master",
    "repository": "cms-repo",
    "authInfo": {
        "principals": [
            "user:system:anonymous",
            "role:system.everyone"
        ]
    }
}
```

Similarly, the getSite() and getSiteConfig() functions in the portal lib also return null. A workaround so far has been to handle the authentication inside the error handler instead of using an idprovider.

**Comment by sigdestad** [comment]
As we talked about Glenn.. This fix should preferably also be added to the 6.12 fix release next week (if possible). FYI @alansemenov

**Comment by GlennRicaud** [comment]
My answer back when we discussed with Bjørnar was that an error handler looked like the correct solution. It was relying on being called under a site and requiring a special site config.
It only needed to handle 401 errors. So it made sense that it was a error handler and not a generic ID Provider (that is supposed to be called from anywhere).

But okay if we need to make ID Provider depend on content context:
To clarify a bit the situation, there are multiple ways the ID Provider can be called:
  1) "autoLogin" invoked only when a user is not authenticated.
  2) "handle401" rendered in case of 401 errors.
  3) All the other methods (login, logout, ...) are similar to services. They are called with "_/[userstore]/method" 

Let's focus on 3) first. The behaviour is the same as a service. So it does actually set the current content and site. But I guess it is not set in this case because the current content/site is forbidden (same behaviour as a service).

Bjørnar is right that there is not enough information in the request passed to the controller.
We have for example a method getContentPath that is not used by the portal request mapper. (I do not know why it is not used).
With that the id provider could have done gotten the site for the 2) and 3) methods with the following code :
```
var site = contextLib.run({
    user: {
        login: 'anonymous',
        userStore: 'system'
    },
    principals: ["role:system.admin"]
}, function(){ return contentLib.getContent(key:req.contentPath);});
```

For handle401, the problem is similar. For example if the problem is a page forbidden, the page handler throw the 401 and of course does not set the content on the portal request.
But the contentPath is set before though. So the previous method would work.

Again it is strange to expect an ID Provider to be launched under a site. I would have personaly used a error handler like bhj.

**Comment by GlennRicaud** [comment]
@tsi: So to sum up. It works like it is supposed to. What is the fix you want to do in 6.12.2?

Or are talking about the config on the virtual host mapping?

**Comment by jsi** [comment]
6.12.3 is through the test phase, and ready for release.  For this issue, it is not even 100% clear what needs to be done.  Therefore, I am pushing the button for 6.12.3 now, then we will continue working on this issue for 6.13.0.

**Comment by sigdestad** [comment]
I just wanted the site context to be available if an idprovider was accessed within a site context?

**Comment by GlennRicaud** [comment]
I close this issue. The site context will be available by the combination of the 2 following issues:
https://github.com/enonic/xp/issues/5854 (Fixed in 6.13)
https://github.com/enonic/xp/issues/5845 (To be implemented in 6.14)
